### PR TITLE
Refactor metric api

### PR DESCRIPTION
**Pull request by 2bys**
This pull request includes a refactoring of the evaluation pipeline, i.e.:
- Add regression metric `crps_gaussian`. 
- Introduce general functions for handling metrics, which can now be used for arbitrary keys in a results dictionary:
  - `apply_fns`: Applies a list of metrics with specific keywords that are extracted from the results of an arbitrary pushforward function. 
  - `transfer_entries`: Function transfers additional results (such as mean, std, ...) from the results of the pushforward functions to the results dictionary of the metric evaluation. 
  - `evaluate_on_generator`: This function takes an arbitrary generator and evaluates pushforward together with metrics in a batched way.
- Updates to the MNIST tutorial outlining how to use these changes - so-called Metrics API.
- Removes unused functions and metrics.

### CODE REVIEW

**Review by bmucsanyi — CHANGES_REQUESTED** [review]

File: laplax/eval/utils.py [hunk 1]
```diff
@@ -137,12 +86,142 @@ def evaluate_data_point(dp: Data) -> dict[str, Array]:
...
+    """
+    # Convert list to dict if needed
+    if isinstance(mapping, list):
+        mapping = {k: k for k in mapping}
+
+    # Transfer the entry
+    def transfer(results, aux, **kwargs):
+        del kwargs
+        options = {"results": results, "aux": aux}
+        if field == "results":
```
[the remark this answers is not in the document]
Comment (2bys) [reply laplax/eval/utils.py:203]: Agree. Will simplify this.

File: examples/0002_laplax_on_mnist.ipynb [hunk 2]
```diff
@@ -519,25 +543,167 @@
...
+    "print(f\"MAP acc: {correctness_map_val.mean():.4f}\")\n",
+    "print(f\"Laplace ECE: {ece_pred:.4f}\")\n",
+    "print(f\"Laplace acc: {correctness_pred_val.mean():.4f}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If all functions (pushforward and metrics) support already a batch dimension, then we can set `has_batch` to `False`, which might be faster in execution."
```
[the remark this answers is not in the document]
Comment (2bys) [reply examples/0002_laplax_on_mnist.ipynb:641]: Will change to: If all functions (pushforward and metrics) support a batch dimension, then we can set `has_batch` to `False`. Otherwise, `has_batch=True` will apply `jax.lax.map` along the batch dimension.

@bmucsanyi Should we keep the previous for-loop evaluation routine?

File: examples/0002_laplax_on_mnist.ipynb [hunk 3]
```diff
@@ -519,25 +543,167 @@
...
+    "## Same evaluation using the Metrics API"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Setup two versions of the pushforward functions.\n",
```
[the remark this answers is not in the document]
Comment (2bys) [reply examples/0002_laplax_on_mnist.ipynb:560]: Agree

File: examples/0002_laplax_on_mnist.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (2bys) [reply examples/0002_laplax_on_mnist.ipynb]: Done

**Review by bmucsanyi — APPROVED** [review]
Thanks, looks good!
